File: ch12_the_sparse_extended_information_filter/src/python/sparsification.py
from numpy import *
from scipy.sparse import csr_matrix
from scipy.linalg import solve

from pdb import set_trace as bp

def sparsification(m0,m1,xi,omega,m,G):
    sg = G.shape[0]
    ol = csr_matrix((ones(m1.size),(zeros(m1.size),m1)),shape=(sg,sg),dtype=bool)
    G = G - ol - ol.T
    G = G.astype(bool)
    s,t = meshgrid(union1d(m0,m1),m1)
    s = s.flatten()
    t = t.flatten()
    edges = sort(vstack((s,t)),axis=0)
    edges = unique(edges,axis=1)   #needs version >=1.13
    #edges = vstack({tuple(row) for row in edges.T}).T
    edges = edges[:,edges[0,:]-edges[1,:]!=0]
    nl = csr_matrix((ones(edges.shape[1]),(edges[0,:],edges[1,:])),shape=(sg,sg),dtype=bool)
    G = G + nl + nl.T

    m0 = sort(m0)
    m1 = sort(m1)
    m0 = 2*m0 + 1
    m0 = vstack((m0,m0+1)).flatten(order='F')
    m1 = 2*m1 + 1
    m1 = vstack((m1,m1+1)).flatten(order='F')

    omega2 = omega[ix_(hstack((hstack((r_[0:3],m0)),m1)),m1)]
    l1 = omega2.dot(solve(omega[ix_(m1,m1)],omega2.T))

    omega3 = omega[ix_(hstack((hstack((r_[0:3],m0)),m1)),hstack((r_[0:3],m1)))]
    l2 = omega3.dot(solve(omega[ix_(hstack((r_[0:3],m1)),hstack((r_[0:3],m1)))],omega3.T))

    omega4 = omega[:,r_[0:3]]
    l3 = omega4.dot(solve(omega[ix_(r_[0:3],r_[0:3])],omega4.T))

    omega1 = copy(omega)
    omega1[ix_(hstack((hstack((r_[0:3],m0)),m1)),hstack((hstack((r_[0:3],m0)),m1)))] -= l1
    omega1[ix_(hstack((hstack((r_[0:3],m0)),m1)),hstack((hstack((r_[0:3],m0)),m1)))] += l2
    omega1 -= l3

    omega1[ix_(r_[0:3],m1)] = 0
    omega1[ix_(m1,r_[0:3])] = 0

    xi1 = xi + (omega1-omega)@m

    return xi1,omega1,G

File: ch12_the_sparse_extended_information_filter/src/julia/sparsification.jl
function sparsification(m₀,m₁,μ,ξ,Ω,graph)

    sg = size(graph,1)
    rows = ones(length(m₁))
    cols = m₁ .+ 1
    adjacency = sparse(rows,cols,true,sg,sg)
    graph .⊻= adjacency .| adjacency'

    edges = [[i+1 j+1] for i ∈ union(m₀,m₁),j ∈ m₁ if i ≠ j]
    edges = vcat(edges...)
    adjacency = sparse(edges[:,1],edges[:,2],true,sg,sg)
    graph .|= adjacency .| adjacency'

    ind₀ = [[2i+2,2i+3] for i ∈ sort(m₀)]
    ind₀ = vcat(ind₀...)
    ind₁ = [[2i+2,2i+3] for i ∈ sort(m₁)]
    ind₁ = vcat(ind₁...)

    Ω₂ = Ω[[1:3;ind₀;ind₁],ind₁]
    l₁ = Ω₂ / Symmetric(Ω[ind₁,ind₁]) * Ω₂'

    Ω₃ = Ω[[1:3;ind₀;ind₁],[1:3;ind₁]]
    l₂ = Ω₃ / Symmetric(Ω[[1:3;ind₁],[1:3;ind₁]]) * Ω₃'

    Ω₄ = Ω[:,1:3]
    l₃ = Ω₄ / Symmetric(Ω[1:3,1:3]) * Ω₄'

    Ω₁ = copy(Ω)
    Ω[[1:3;ind₀;ind₁],[1:3;ind₀;ind₁]] .-= l₁
    Ω[[1:3;ind₀;ind₁],[1:3;ind₀;ind₁]] .+= l₂
    Ω .-= l₃

    ξ .+= (Ω-Ω₁)*μ
end
